### DISCUSSION TITLE
Beginner Help with DS18B20 Config

### DISCUSSION THREAD

**Discussion by pbix** [arendst/Tasmota#18230] [Q&A]
Hi Folks,
New to Tasmota here.  I purchased a Espressif ESP32-S2-DevKitM1.  I loaded it with [tasmota32s2cdc.bin](http://ota.tasmota.com/tasmota32/release-12.2.0/tasmota32s2cdc.bin)

Then after getting it connected to my WIFI I selected Configuration -> Configure Other and applied the template I found here.
[https://templates.blakadder.com/espressif_ESP32-S2-DevKitM-1.html](url)

I SetOption129 On to enable display of the ESP32 temperature.  At this point the UI looks proper.  (See screenshot below)

Finally, I have a DS18B20 which I connected to 3.3V to ground and the yellow data wire to GPIO9 with 4.7k pull-up resistor to V3.3.

Under Configuration -> Configure Module I selected DS18x20 on this GPIO pin.  Then after saving configuration the device reboots but no sensor appears on the UI.  

I cannot seen any evidence in the console that even looked for the device at boot time.

Any help in correcting my procedure would be greatly appreciated.

`00:00:00.002 CMD: Using USB CDC
00:00:00.002 HDW: ESP32-S2FH4 
00:00:00.007 UFS: FlashFS mounted with 304 kB free
00:00:00.019 CFG: Loaded from File, Count 14
00:00:00.024 QPC: Count 1
00:00:00.091 BRY: Berry initialized, RAM used=4068 bytes
00:00:00.116 Project tasmota - Tasmota Version 12.2.0(tasmota)-2_0_5(2022-10-17T08:40:02)
00:00:00.824 WIF: Connecting to AP1 BDeviceNet Channel 11 BSSId 6A:FF:7B:F9:BF:03 in mode 11n as tasmota-D9E1E4-0484...
00:00:03.247 WIF: Connected
00:00:03.508 HTP: Web server active on tasmota-D9E1E4-0484 with IP address 192.168.2.218
01:42:16.275 RSL: INFO1 = {"Info1":{"Module":"ESP32-S2 DevKitM-1","Version":"12.2.0(tasmota)","FallbackTopic":"cmnd/DVES_D9E1E4_fb/","GroupTopic":"cmnd/tasmotas/"}}
01:42:16.278 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-D9E1E4-0484","IPAddress":"192.168.2.218"}}
01:42:16.280 RSL: INFO3 = {"Info3":{"RestartReason":"Software reset digital core","BootCount":6}}
01:42:16.282 RSL: RESULT = {"POWER":"ON"}
01:42:16.283 RSL: POWER = ON
01:42:19.014 QPC: Reset
01:42:21.019 RSL: STATE = {"Time":"2023-03-21T01:42:21","Uptime":"0T00:00:10","UptimeSec":10,"Heap":75,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":101,"MqttCount":0,"Berry":{"HeapUsed":3,"Objects":44},"POWER":"ON","Dimmer":10,"Color":"1A1A1A","HSBColor":"0,0,10","Channel":[10,10,10],"Scheme":0,"Width":1,"Fade":"OFF","Speed":1,"LedTable":"ON","Wifi":{"AP":1,"SSId":"BixelDeviceNet","BSSId":"6A:FF:7B:F9:BF:03","Channel":11,"Mode":"11n","RSSI":94,"Signal":-53,"LinkCount":1,"Downtime":"0T00:00:04"}}
01:47:21.024 RSL: STATE = {"Time":"2023-03-21T01:47:21","Uptime":"0T00:05:10","UptimeSec":310,"Heap":72,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":99,"MqttCount":0,"Berry":{"HeapUsed":3,"Objects":44},"POWER":"ON","Dimmer":10,"Color":"1A1A1A","HSBColor":"0,0,10","Channel":[10,10,10],"Scheme":0,"Width":1,"Fade":"OFF","Speed":1,"LedTable":"ON","Wifi":{"AP":1,"SSId":"BixelDeviceNet","BSSId":"6A:FF:7B:F9:BF:03","Channel":11,"Mode":"11n","RSSI":94,"Signal":-53,"LinkCount":1,"Downtime":"0T00:00:04"}}
01:52:11.010 APP: Serial logging disabled
01:52:21.014 RSL: STATE = {"Time":"2023-03-21T01:52:21","Uptime":"0T00:10:10","UptimeSec":610,"Heap":76,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":99,"MqttCount":0,"Berry":{"HeapUsed":3,"Objects":44},"POWER":"ON","Dimmer":10,"Color":"1A1A1A","HSBColor":"0,0,10","Channel":[10,10,10],"Scheme":0,"Width":1,"Fade":"OFF","Speed":1,"LedTable":"ON","Wifi":{"AP":1,"SSId":"BixelDeviceNet","BSSId":"6A:FF:7B:F9:BF:03","Channel":11,"Mode":"11n","RSSI":94,"Signal":-53,"LinkCount":1,"Downtime":"0T00:00:04"}}v`

![Screenshot_12_1_1](https://user-images.githubusercontent.com/289729/226496314-12f18c3b-6a81-4835-932e-6b3b1c12868c.jpg)

**Comment by barbudor** [comment]
Try command 'gpio all' in the console
The answer should confirm that the GPIO you Co figured for DS18X20 is actually properly seen as it

There are no many people using Tasmota of S2 so may be we are missing some feedback. The driver is different for esp8266 and esp32 but I don't know to which extend it has been tested on S2. 

@Jason2866 did you ever tested DS18X20 on S2 ?

**Reply by sfromis** [reply]
Did a test on a Wemos S2 Mini board, with a DS18B20 sensor (including pullup resistor) on gpio 9, and was able to get it to work normally. I did have a bit of trouble, which I guess to be due to dubious dupont connections.

With `Weblog 3` activated, at boot time I see `DSB: Sensors found 1` to indicate that the sensor was detected, or not if the value 0 is displayed instead.

**Reply by Jason2866** [reply]
Does work. Tested with latest dev, GPIO 9 used
```
tele/sonoff-0C0A3C/SENSOR = {"Time":"2023-03-23T10:50:55","DS18B20":{"Id":"0114539A22FF","Temperature":21.4},"ESP32":{"Temperature":38.5},"TempUnit":"C"}
```

**Comment by pbix** [comment]
Thanks for the suggestions.  Results from the "gpio all" command below.  I  think I see the sensor is detected on GPIO 9?  What else can I try?  Its still not showing on the main screen.

```00:48:00.312 CMD: gpio all
00:48:00.319 RSL: RESULT = {"GPIO0":{"32":"Button1"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"0":"None"},"GPIO4":{"0":"None"},"GPIO5":{"0":"None"},"GPIO6":{"0":"None"},"GPIO7":{"0":"None"},"GPIO8":{"0":"None"},"GPIO9":{"1312":"DS18x20"},"GPIO10":{"0":"None"},"GPIO11":{"0":"None"},"GPIO12":{"0":"None"},"GPIO13":{"0":"None"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"},"GPIO18":{"1376":"WS28121"},"GPIO19":{"0":"None"},"GPIO20":{"0":"None"},"GPIO21":{"0":"None"},"GPIO33":{"0":"None"},"GPIO34":{"0":"None"},"GPIO35":{"0":"None"},"GPIO36":{"0":"None"},"GPIO37":{"0":"None"},"GPIO38":{"0":"None"},"GPIO39":{"0":"None"},"GPIO40":{"0":"None"},"GPIO41":{"0":"None"},"GPIO42":{"0":"None"},"GPIO43":{"0":"None"},"GPIO44":{"0":"None"},"GPIO45":{"0":"None"},"GPIO46":{"0":"None"}}```

Results of weblog 3 after restart:
```00:00:00.002 CMD: Using USB CDC
00:00:00.002 HDW: ESP32-S2FH4 
00:00:00.003 HDW: FoundPSRAM=0 CanUsePSRAM=0
00:00:00.015 UFS: FlashFS mounted with 304 kB free
00:00:00.032 CFG: Loaded from File, Count 16
00:00:00.037 QPC: Count 1
00:00:00.038 CFG: CR 390/699, Busy 0
00:00:00.045 ROT: Mode 1
00:00:00.120 BRY: Berry initialized, RAM used=4068 bytes
00:00:00.146 BRY: No 'preinit.be'
00:00:00.148 SRC: Restart
00:00:00.154 Project tasmota - Tasmota Version 12.2.0(tasmota)-2_0_5(2022-10-17T08:40:02)
00:00:00.171 DSB: Sensors found 0
00:00:00.256 BRY: No 'autoexec.be'
00:00:00.515 WIF: Attempting connection...
00:00:00.893 WIF: Connecting to AP1 BixelDeviceNet Channel 11 BSSId 6A:FF:7B:F9:BF:03 in mode 11n as tasmota-D9E1E4-0484...
00:00:03.236 WIF: Connected
00:00:03.495 HTP: Web server active on tasmota-D9E1E4-0484 with IP address 192.168.2.218
00:00:04.475 RTC: UTC 2023-03-21T23:52:09, DST 2023-03-26T02:00:00, STD 2023-10-29T03:00:00
00:52:09.000 RTC: Synced by NTP
00:52:09.271 RSL: INFO1 = {"Info1":{"Module":"ESP32-S2 DevKitM-1","Version":"12.2.0(tasmota)","FallbackTopic":"cmnd/DVES_D9E1E4_fb/","GroupTopic":"cmnd/tasmotas/"}}
00:52:09.273 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-D9E1E4-0484","IPAddress":"192.168.2.218"}}
00:52:09.276 RSL: INFO3 = {"Info3":{"RestartReason":"Software reset digital core","BootCount":7}}
00:52:09.278 RSL: RESULT = {"POWER":"ON"}
00:52:09.279 RSL: POWER = ON
00:52:09.282 TFS: File 'autoexec.bat' not found
00:52:10.917 HTP: Main Menu
00:52:11.068 QPC: Reset
00:52:12.961 HTP: Consoles
00:52:13.067 APP: Boot Count 7
00:52:13.372 CFG: Saved, Count 17, Bytes 4096
00:52:14.068 RSL: STATE = {"Time":"2023-03-22T00:52:14","Uptime":"0T00:00:11","UptimeSec":11,"Heap":75,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":99,"MqttCount":0,"Berry":{"HeapUsed":3,"Objects":44},"POWER":"ON","Dimmer":10,"Color":"1A1A1A","HSBColor":"0,0,10","Channel":[10,10,10],"Scheme":0,"Width":1,"Fade":"OFF","Speed":1,"LedTable":"ON","Wifi":{"AP":1,"SSId":"BixelDeviceNet","BSSId":"6A:FF:7B:F9:BF:03","Channel":11,"Mode":"11n","RSSI":98,"Signal":-51,"LinkCount":1,"Downtime":"0T00:00:05"}}
00:52:15.011 HTP: Console```

**Reply by sfromis** [reply]
`DSB: Sensors found 0`

Tasmota is doing what it can. Now it is up to the hardware to do better. Do verify your wiring carefully.
You may be a victim of: https://github.com/cpetrich/counterfeit_DS18B20/

BTW. To post lines in a readable format, wrap them in triple backtick before and after, instead of only single backtick.

**Comment by pbix** [comment]
I guess I was thinking the UI would show a sensor with an invalid reading if there was a hardware problem.

Are you saying that even if I configure the I/O the UI will not show anything unless the sensor is detected at boot time?

Finally, what is the difference between "Configure Module" and "Configure Template"?  These two seem to have the same function to a newbie like me.  How do I know which one to use?

**Reply by sfromis** [reply]
Configure Module means that devices need to be configured pin by pin. Using a template means that the whole pin config (and a bit more) can be set with one copy/paste, and then the template can be modified (or created) in "Configure Template". When the template is activated (else not used), "Configure Module" only has pins left as "User" in the template.

> Are you saying that even if I configure the I/O the UI will not show anything unless the sensor is detected at boot time?

Yes. You get the boot time message in the log if/when detected.

**Comment by pbix** [comment]
I can say I have had these probes working on this same board running ESPEasy.  I also I have these probes running on a Shelly Uni so I am a bit skeptical that it is the probes.

I did read the link regarding counterfeits.  I do not remember if I got these from Digikey or Amazon.  But it seems to me the article describes accuracy issues and not so much that it doesn't work at all.

I will put a scope on it this weekend and see if I can see any activity on the data line.

**Comment by Jason2866** [comment]
@pbix Please use latest release or development version. Our tests are always done with actual versions.
Your version is not up to date `Tasmota Version 12.2.0(tasmota)-2_0_5(2022-10-17T08:40:02`

**Reply by pbix** [reply]
I apologize that I did not note above that I have tested this with the current release and the previous four releases.  
As far as I can determine these are the only releases that support ESP32-S2-DevKit.

I have not tested with the bleeding edge.

**Comment by pbix** [comment]
I am concerned there may be some confusion about which GPIO pin is mapped where.  Is there a way for Tasmota to check every GPIO pin for this device so that I can confirm that GPIO9 is the correct mapping in my case.

There is another observation I want to report.  When I initially loaded the current release of Tasmota on my board, when I selected DS18x20 for my pin there was a new field that appeared right beside the selection that allowed me to select 1-4.  I was not sure what this was for but I always selected one.  But since trying other older versions of Tasmota I have observed that this field no longer appears.  So it leaves me concerned that something is wrong in the config somehow.  Does anyone know what this field was intended to indicate and why it no longer appears?

**Reply by sfromis** [reply]
It is fine leaving the dropdown appearing to the right on the default of one, as this is for the case of having DS18x20 on more than one pin. As this is a relatively new feature, it will not appear on previous versions.

One way of verifying that the Tasmota GPIO9 is what you connect to is config it as an input pin, like Button/Switch/Counter, and see if you get reaction when connecting the pin to ground. Or config it as a relay, and see if a connected led reacts.

**Reply by pbix** [reply]
To be clear, the drop down no longer appears on any version, this is my observation.  Once I downgraded it disappeared and never came back after I upgraded back to the current release.

Thanks for your tip on switching to input.  I will try it.

**Reply by Jason2866** [reply]
Seems something is "broken". Install again with Tasmota Web Installer and set a tick in the ease option box.

**Reply by sfromis** [reply]
Downgrading is not a supported operation.

**Reply by pbix** [reply]
Upgrade back to v12.4.0, was able to reset the configuration from the initial Wifi connect screen.  Re-applied the DevKit template so everything should be back to defaults.  The drop down menu appears again.

Switching to inputs/switch/button as suggested by @sfromis does not cause anything to appear on the main menu UI so no way to tell if the GPIO is working or not.  Again I am a new user, what input type do I select so that something appears on the Main Menu UI so that I can tell the status of the mapped input?

**Reply by barbudor** [reply]
Actually no buttons or switch inputs are reported in the front page but toggling the GPIO would trigger some messages in the console

Alternatively if you set 1 gpio as switch_1 and another as Relay_1 toggling the switch gpio would automatically toggle the associated relay gpio and that would be reported on the gui

**Reply by sfromis** [reply]
Easiest may be to config the input pin as a Counter, and then see if the number appearing on the front page increments when tickling the pin.

**Comment by pbix** [comment]
In followup on this post.  I cannot get any inputs to work with this device.  As noted above not too many use it so it could be a bug or as I am a newbie to Tasmota it could be user error.  But I would caution anyone contemplating purchasing the Espressif ESP32-S2-DevKitM1 that you may have issues.

I am looking now for advice for what ESP32 device I should purchase this is known to work well.  I need one that can pin mount like this board does and has at least 14 GPIO.
